src/hb-ot-cmap-table.hh

```diff
@@ -1468,7 +1503,155 @@ struct cmap
     | hb_filter ([&] (const hb_pair_t<hb_codepoint_t, hb_codepoint_t> _)
 		 { return (_.second != HB_MAP_VALUE_INVALID); })
     ;
-    cmap_prime->serialize (c->serializer, it, encodingrec_iter, this, c->plan);
+
+    for (auto _ : format12_tables.iter ())
+    {
+      unsigned language =_.second;
+      const EncodingRecord *format4_tbl = nullptr;
+      const EncodingRecord *format12_tbl = hb_addressof (encodingRecord[_.first]);
+      if (format12_tbl->platformID == 0 && format12_tbl->encodingID == 4)
+      {
+        if (!unicode_bmp_tables.has (language))
+        {
+          format12_tables.del (_.first);
+          continue;
+        }
+        format4_tbl = unicode_bmp_tables.get (language);
+      }
+
+      if (format12_tbl->platformID == 3 && format12_tbl->encodingID == 10)
+      {
+        if (!ms_bmp_tables.has (language))
+        {
+          format12_tables.del (_.first);
+          continue;
+        }
+        format4_tbl = ms_bmp_tables.get (language);
+      }
+
+      if (!format4_tbl) continue;
+
+      if (!c->plan->glyphs_requested->is_empty ())
+      {
+        hb_set_t format12_unicodes_set;
+        hb_map_t format12_cp_glyphid_map;
+        (this+format12_tbl->subtable).collect_mapping (&format12_unicodes_set, &format12_cp_glyphid_map);
+
+        auto format12_iter =
+        + hb_zip (format12_unicodes_set.iter(), format12_unicodes_set.iter() | hb_map(format12_cp_glyphid_map))
+        | hb_filter (c->plan->_glyphset, hb_second)
+        | hb_filter ([c] (const hb_pair_t<hb_codepoint_t, hb_codepoint_t>& p)
+                     {
+                       return c->plan->unicodes->has (p.first) ||
+                              c->plan->glyphs_requested->has (p.second);
+                     })
+        | hb_map (hb_first)
```

Comment:
Add a helper method which creates this iterator. It could then be reused for the format4 table and again in the serialize method which I believe has the exact same iterator